packages/svelte/src/compiler/phases/2-analyze/css/css-prune.js

```diff
@@ -1,261 +1,387 @@
+			case ' ':
+			case '>': {
+				let parent = /** @type {import('#compiler').TemplateNode | null} */ (element.parent);
 
-	if (relative_selector.combinator) {
-		if (
-			relative_selector.combinator.type === 'Combinator' &&
-			relative_selector.combinator.name === ' '
-		) {
-			for (const ancestor_selector of relative_selectors) {
-				if (ancestor_selector.metadata.is_global) {
-					continue;
-				}
+				let parent_matched = false;
+				let crossed_component_boundary = false;
 
-				if (ancestor_selector.metadata.is_host) {
-					return mark(relative_selector, element);
-				}
+				while (parent) {
+					if (parent.type === 'Component' || parent.type === 'SvelteComponent') {
+						crossed_component_boundary = true;
+					}
 
-				/** @type {import('#compiler').RegularElement | import('#compiler').SvelteElement | null} */
-				let parent = element;
-				let matched = false;
-				while ((parent = get_element_parent(parent))) {
-					if (relative_selector_might_apply_to_node(ancestor_selector, parent) !== NO_MATCH) {
-						mark(ancestor_selector, parent);
-						matched = true;
+					if (parent.type === 'RegularElement' || parent.type === 'SvelteElement') {
+						if (apply_selector(parent_selectors, rule, parent, stylesheet)) {
+							// TODO the `name === ' '` causes false positives, but removing it causes false negatives...
+							if (name === ' ' || crossed_component_boundary) {
+								mark(parent_selectors[parent_selectors.length - 1], parent);
+							}
+
+							parent_matched = true;
+						}
+
+						if (name === '>') return parent_matched;
 					}
-				}
 
-				if (matched) {
-					return mark(relative_selector, element);
+					parent = /** @type {import('#compiler').TemplateNode | null} */ (parent.parent);
 				}
-			}
 
-			if (relative_selectors.every((relative_selector) => relative_selector.metadata.is_global)) {
-				return mark(relative_selector, element);
+				return parent_matched || parent_selectors.every((selector) => is_global(selector, rule));
 			}
 
-			return false;
-		}
+			case '+':
+			case '~': {
+				const siblings = get_possible_element_siblings(element, name === '+');
 
-		if (relative_selector.combinator.name === '>') {
-			const has_global_parent = relative_selectors.every(
-				(relative_selector) => relative_selector.metadata.is_global
-			);
+				let sibling_matched = false;
 
-			if (
-				has_global_parent ||
-				apply_selector(relative_selectors, get_element_parent(element), stylesheet)
-			) {
-				return mark(relative_selector, element);
+				for (const possible_sibling of siblings.keys()) {
+					if (apply_selector(parent_selectors, rule, possible_sibling, stylesheet)) {
+						mark(relative_selector, element);
+						sibling_matched = true;
+					}
+				}
+
+				return (
+					sibling_matched ||
+					(get_element_parent(element) === null &&
+						parent_selectors.every((selector) => is_global(selector, rule)))
+				);
 			}
 
-			return false;
+			default:
+				// TODO other combinators
```

Comment:
Isn't it better to throw an error if we're missing a combinator?

Reply:
No, it's better to just be conservative and include the CSS — removing it is an optimisation. The only combinator we're missing is `||`. If another combinator is added to the language, all we need to do (until we have time to implement pruning for it, which may be prohibitively difficult) is adjust a regex